### PR TITLE
Sub organization implementation

### PR DESCRIPTION
**Pull request by hari-hypto**
Approach details - https://www.notion.so/togai/Authn-and-Authz-design-for-customer-and-reseller-portal-af64664bff4c496480ad88eb1040bf56?pvs=4

### CODE REVIEW

File: src/main/kotlin/com/hypto/iam/server/service/UsersService.kt [hunk 1]
```diff
@@ -387,27 +426,46 @@ class UsersServiceImpl : KoinComponent, UsersService {
...
-        identityProvider.authenticate(identityGroup, userName, password)
+        val encodedEmail = if (subOrganizationId != null) {
+            encodeSubOrgUserEmail(
+                email,
+                organizationId,
+                subOrganizationId
+            )
+        } else {
+            email
+        }
```
Comment (karthick-vinod) [inline src/main/kotlin/com/hypto/iam/server/service/UsersService.kt:458]: nit: Code duplication. Can be moved to a separate function.

File: src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt [hunk 2]
```diff
@@ -133,6 +134,11 @@ fun CreateUserRequest.validate(): CreateUserRequest {
     return createUserRequestValidation.validateAndThrowOnFailure(this)
 }
 
+fun CreateSubOrganizationRequest.validate(): CreateSubOrganizationRequest {
+    // TODO: Add validation for sub organization
```
Comment (karthick-vinod) [inline src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt:138]: Address this TODO

**Review by karthick-vinod — APPROVED** [review]
- How are we passing the sub organization information in JWT token? Will the sub claim contain encoded email or actual email? FE might use that information. Also, please update JWT.md with the same information.
- Should we add the ability to update / customize audience of the token? `aud` claim for customer and reseller portal use cases

File: iam_openapi_spec.yml [hunk 3]
```diff
@@ -189,6 +189,134 @@ paths:
...
+        '403':
+          $ref: '#/components/responses/ErrorResponse'
+        '404':
+          $ref: '#/components/responses/ErrorResponse'
+        '429':
+          $ref: '#/components/responses/ErrorResponse'
+        default:
+          $ref: '#/components/responses/ErrorResponse'
+
+  /organizations/{organization_id}/sub_organizations/{sub_organization_id}:
```
Comment (karthick-vinod) [inline iam_openapi_spec.yml:244]: /organizations/{organization_id}/sub_organizations/{sub_organization_id}/users/{user_id}

APIs are not present in openapi spec

**Comment by hari-hypto** [comment]
> * How are we passing the sub organization information in JWT token? Will the sub claim contain encoded email or actual email? FE might use that information. Also, please update JWT.md with the same information.

We get the sub org information from userHrn. JWT token contains only normal email address but today the sub org information is not explicitly passed in JWT token. There is no change actually in JWT.md probably we can just mention that user hrn now can contain sub org info.

> * Should we add the ability to update / customize audience of the token? `aud` claim for customer and reseller portal use cases

I dont know whether its needed today. 'aud' definitely gives more fine grained permissions control but it comes with extra overhead of support to vend tokens for certain 'aud' and validate them at the resource server side. Today with JWT policies this is partially solved. (but definitely not the best security practice). We can later support 'aud' in our tokens as JWT tokens are max 1 day valid. We will not have big backward compatibility issue when we want to support 'aud'.

File: iam_openapi_spec.yml [hunk 4]
```diff
@@ -2237,6 +2457,27 @@ components:
...
+          name: ACME Enterprise - Customer 123
+          description: Brief description about ACME
+          updatedAt: 2020-01-01T00:00:00.000Z
+          createdAt: 2020-01-01T00:00:00.000Z
+
+    SubOrganizationsPaginatedResponse:
+      value:
+        data:
+          - id: XlK5h64gND
+            name: invoice
```
[the remark this answers is not in the document]
Comment (hari-hypto) [reply iam_openapi_spec.yml:2473]: Done

File: src/main/kotlin/com/hypto/iam/server/security/Authorization.kt [hunk 5]
```diff
@@ -171,15 +172,7 @@ fun getResourceHrnFunc(
...
-data class HrnTemplateInput(
-    val pathTemplate: String,
-    val resourceNameIndex: Int,
-    val resourceInstanceIndex: Int,
-    val organizationIdIndex: Int,
-    val subOrganizationIdIndex: Int? = null,
-)
-
-fun getResourceHrnFunc(templateInputs: List<HrnTemplateInput>): (ApplicationRequest) -> ResourceHrn {
+fun getResourceHrnFunc(templateInputs: List<RouteOption>): (ApplicationRequest) -> ResourceHrn {
```
Comment (karthick-vinod) [inline src/main/kotlin/com/hypto/iam/server/security/Authorization.kt:175]: In the new model, are we passing `List<RouteOption>` anywhere? If not, we can make this function accept just one element.

**Review by karthick-vinod — APPROVED** [review]

File: iam_openapi_spec.yml [hunk 6]
```diff
@@ -530,6 +713,36 @@ paths:
       security:
         - apiKeyAuth: []
 
+
+  /organizations/{organization_id}/sub_organizations/{sub_organization_id}/users/resetPassword:
```
Comment (karthick-vinod) [inline iam_openapi_spec.yml]:
```suggestion
  /organizations/{organization_id}/sub_organizations/{sub_organization_id}/users/{user_name}/resetPassword:
```

File: src/main/kotlin/com/hypto/iam/server/service/UsersService.kt [hunk 7]
```diff
@@ -447,20 +447,26 @@ class UsersServiceImpl : KoinComponent, UsersService {
...
+        email: String,
+    ) = if (subOrganizationName != null) {
+        encodeSubOrgUserEmail(
+            email,
+            organizationId,
+            subOrganizationName
+        )
+    } else {
+        email
+    }
```
Comment (karthick-vinod) [inline src/main/kotlin/com/hypto/iam/server/service/UsersService.kt:468]:
```suggestion
    ) = subOrganizationName?.let {
      encodeSubOrgUserEmail(it, organizationId, subOrganizationName)
    } ?: email

```

File: src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt (hunk 2 above)
Comment (karthick-vinod) [reply src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt:139]: Remove this TODO comment

File: src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt [hunk 8]
```diff
@@ -449,6 +450,17 @@ val updateUserRequestValidation = Validation {
     }
 }
 
+val createSubOrganizationRequest = Validation {
+    CreateSubOrganizationRequest::name required {
+        maxLength(MAX_SUB_ORG_LENGTH) hint "Maximum length supported for" +
+            "name is $MAX_SUB_ORG_LENGTH characters"
+        noEndSpaces()
```
Comment (karthick-vinod) [inline src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt:457]: Will this validate or mutate the input string?

**Review by karthick-vinod — APPROVED** [review]

File: iam_openapi_spec.yml (hunk 6 above)
Comment (hari-hypto) [reply iam_openapi_spec.yml]: Resolving as this is not required. This API is triggered during forgot password flow.

File: src/main/kotlin/com/hypto/iam/server/service/UsersService.kt (hunk 7 above)
Comment (hari-hypto) [reply src/main/kotlin/com/hypto/iam/server/service/UsersService.kt:468]: Addressed in this PR - https://github.com/hwslabs/iam/pull/163

File: src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt (hunk 8 above)
Comment (hari-hypto) [reply src/main/kotlin/com/hypto/iam/server/validators/RequestModelValidator.kt:457]: noEndSpaces() should not be used. Updated in this PR https://github.com/hwslabs/iam/pull/163

**Pull request merged**